accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+HttpClient client = new HttpClient(new HttpClientHandler());
+```
+
+There are several reasons why using `SocketHttpHandler` is not desirable or not even possible on some platforms
+
+* Lack of platform native TLS support for better performance
+* Bloats the application size
+* Performance and power consumption impact of managed implementation
+* Integration with the platform settings
+* Unavailability of Socket APIs
+
+### Platform-specific Http properties need to be available for customization
+In cases where HTTP handler wraps platform-specific implementation and the implementation has a way to control HTTP connection behaviour or properties, such properties need to be exposed to C# developers on the public platform-specific type.  If HttpClientHandler itself exposes such a knob, it needs to control the associated customization on the underlying platform.
+
+A few examples of platform handlers control properties
+
+* [cachePolicy](https://developer.apple.com/documentation/foundation/nsurlrequest/cachepolicy/useprotocolcachepolicy)
+* [allowsConstrainedNetworkAccess](https://developer.apple.com/documentation/foundation/nsmutableurlrequest/3325676-allowsconstrainednetworkaccess)
+* [allowsExpensiveNetworkAccess](https://developer.apple.com/documentation/foundation/nsmutableurlrequest/3325677-allowsexpensivenetworkaccess)
+* [networkServiceType](https://developer.apple.com/documentation/foundation/nsmutableurlrequest/1412378-networkservicetype)
+* [ReadTimeout](https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#setReadTimeout(int))
+* [AllowUserInteraction](https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#setAllowUserInteraction(boolean))
+
+## Xamarin.iOS/tvOS/watchOS
+There are 3 handlers available where **NSUrlSession** handler is the default. Except for watchOS where only NSUrlSessionHandler is supported.
+
+The handler is only preselected as the default with [UI options available](https://docs.microsoft.com/en-us/xamarin/cross-platform/macios/http-stack#selecting-an-httpclient-stack) using VS Project Settings dialog to change them. This is required to “fix” existing HttpClient code which was written with the assumption that the default handler will be the one working best on the platform.
+
+The ILLinker is used to remove the unnecessary handlers from the final binary in this case.
+
+### Handlers in Detail
+**SocketsHttpHandler Handler**
+
+The handler which is purely managed and behaves generally the same across all platforms except WatchOS (sockets are not available). The downside of using it is mostly felt in performance compared to the native handlers.
+
+**CFNetwork Handler**
+
+The CFNetwork handler is based in the old Network Stack API provided by Apple. This handler, while it has better performance, lacks features compared with the other ones. It is mostly around for legacy reasons such as:
+
+* Support old servers that do not work with NSUrlSessionHandler
```

Comment:
I don't recall server incompatibilities, maybe SSLv2 ?

`CFNetworkHandler` mostly exists because `NSUrlSession` API came later (and it was too late to remove without breaking code).

Reply:
Removed the text